### PR TITLE
Add platform metrics logging and event logging

### LINKED ISSUE #4299: Periodic logging of key platform metrics
## Feature Suggestion

Sensu Go already provides a /metrics API, a Prometheus exporter that makes certain Sensu performance metrics available in Prometheus exposition format. 

It would be helpful if some ~~or all~~ of these same metrics were emitted via a log file (e.g. stats.log) for evaluation and troubleshooting without external tooling.

## Possible Implementation

~Proposed key metrics (I'm sure I'm missing some):~

~- **events/s**~
~- **bytes/s** (see https://github.com/sensu/sensu-go/issues/4298)~
~- **buffered event stats** (quantity and perhaps also bytes?)~

Please note the following requirements, which are the result of our [6/28 meeting](https://docs.google.com/document/d/1ozfPtJDD9ySbikRKZJT7OXDxz3Oeata0-4Zum5kvZzc/edit): 

Requirements: 

1. New `--disable-platform-metrics` flag (i.e. making platform metrics logging an opt-out)

1. New `--platform-metrics-logging-interval` flag with a default value of `60` (seconds)

1. New `--platform-metrics-log-file` flag with a default value of `< --state-dir >/stats.log`

1. New `--platform-metrics-format` flag with a default value of `influxdb_line`

1. Use the existing Unix signal for log rotation reopen, copy logic used in the event logger feature

1. List of metrics to log

   Three major components to collect metrics from: WizardBus, Agents per backend, Etcd (disk sync speed, memory usage, etc). 

   **WizardBus Metrics**

   - `sensu_go_wizard_bus{topic="sensu:event-raw"}` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L2700))
   - `sensu_go_wizard_bus{topic="sensu:event"}` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L2707))
   - `sensu_go_wizard_bus{topic="sensu:keepalive"}` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L2714))
   - `sensu_go_wizard_bus{topic="sensu:check"}` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L2721))
   - `sensu_go_wizard_bus{topic="sensu:entity-config:default:sensu-centos"}` 

   **Backend Metrics**
   
   - `sensu_go_event_handler_duration` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L81))
   - `sensu_go_events_processed` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L175))
   - `sensu_go_events_processed{status="success"}` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L270))
   - `sensu_go_events_processed{status="failure"}` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L277))
   - `sensu_go_agent_sessions`
   
   **Etcd Metrics**

   - `etcd_debugging_mvcc_keys_total` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L371))
   - `etcd_debugging_mvcc_delete_total` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L465))
   - `etcd_debugging_mvcc_put_total` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L473))
   - `etcd_debugging_mvcc_range_total` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L481))
   - `etcd_debugging_mvcc_txn_total` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L489))
   - `etcd_debugging_mvcc_db_total_size_in_bytes` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L587))
   - `etcd_disk_wal_fsync_duration_seconds_bucket` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L687))
   - `etcd_disk_backend_commit_duration_seconds_bucket` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L697))
   - `etcd_network_client_grpc_received_bytes_total` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L1627))
   - `etcd_network_client_grpc_sent_bytes_total` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L1726))
   - `etcd_network_peer_received_bytes_total` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L1825))
   - `etcd_network_peer_sent_bytes_total` ([source](https://github.com/sensu/sensu-perf/blob/main/grafana/Sensu_Benchmarks-1562276016486.json#L1926))
   - `etcd_snap_db_fsync_duration_seconds_bucket`

### PR DESCRIPTION
**Pull request by echlebek**
## What is this change?

This commit adds platform metrics logging to sensu-go. The feature
defaults to being enabled, but can be disabled via flag, config file, or
environment variable. See documentation for details.

This commit also adds event logging, previously an enterprise-only
feature, to OSS sensu. The flag names and behaviours for this feature
are the same as in the enterprise product.

Both of these logs can be rotated by sending SIGHUP to sensu-backend
after mv-ing the log files.

## Why is this change necessary?

Closes #4299

## Does your change need a Changelog entry?

Yes

## Were there any complications while making this change?

I had to do some significant refactoring to send one signal to two different components. I also moved the enterprise-only event logging into OSS (with leadership's blessing).

## Have you reviewed and updated the documentation for this change? Is new documentation required?

@hillaryfraley documentation updates will be needed, as there are several new flags available. Please schedule some time with me to go over the changes before we release 6.5, if you feel a demo/explanation is needed. :)

## How did you verify this change?

In addition to unit testing I manually ran the backend and verified that logging was working when configured, and rotation worked as expected.

## Is this change a patch?

Nein!

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -28,6 +28,12 @@ agent processed a particular event.
 - Added `pipelines` field to `Check` and `CheckConfig`.
 - Added `sensu_go_agentd_event_bytes` & `sensu_go_store_event_bytes` summary
 metrics to the `/metrics` endpoint.
+- Added the platform metrics log. This log contains a listing of core Sensu
+metrics in influx-line format. It is enabled by default but can be disabled
+with the --disable-platform-metrics flag. By default the log is appended to
+every 60s, and written to /var/lib/sensu/sensu-backend/stats.log.
```
Comment (ccressent) [inline CHANGELOG.md:34]: Why `/var/lib` and not `/var/log`?

File: backend/cmd/start.go [hunk 2]
```diff
@@ -532,6 +569,22 @@ func flagSet(server bool) *pflag.FlagSet {
...
+
+		// Use a default value of 100,000 messages for the buffer. A serialized event
+		// takes a minimum of around 1300 bytes, so once full the buffer ring could
+		// require about 130MB of memory.
+		_ = flagSet.Int(flagEventLogBufferSize, 100000, "buffer size of the event logger")
+
+		// Use a default value of 10ms for the full buffer wait time. When the buffer
+		// is full, the logger will wait for the writer to consume events from the buffer.
+		// This helps reduce event data loss but comes at the cost of event back-pressure
+		// for the backend and it's agent sessions. If the buffer fills and the wait time
```
Comment (ccressent) [inline backend/cmd/start.go:583]: s/it's/its

File: backend/logging/logrus.go [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package logging
+
+import "github.com/sirupsen/logrus"
+
+var logger = logrus.WithFields(logrus.Fields{
+	"component": "backend",
```
Comment (ccressent) [inline backend/logging/logrus.go:6]: Is that the right component name?

**Review by ccressent — COMMENTED** [review]
Looks excellent to me, just a few nitpick comments/questions

File: CHANGELOG.md (hunk 1 above)
Comment (echlebek) [reply CHANGELOG.md:34]: Good point! It should probably be /var/log.

File: backend/logging/logrus.go (hunk 3 above)
Comment (echlebek) [reply backend/logging/logrus.go:6]: I think so... I didn't want to name the component "logging" :D

**Review by ccressent — APPROVED** [review]

**Pull request merged**